vhost-device-vsock/src/main.rs

```diff
@@ -266,20 +269,32 @@ pub(crate) fn start_backend_server(
 
 pub(crate) fn start_backend_servers(configs: &[VsockConfig]) -> Result<(), BackendError> {
     let cid_map: Arc<RwLock<CidMap>> = Arc::new(RwLock::new(HashMap::new()));
-    let mut handles = Vec::new();
+    let mut handles = HashMap::new();
+    let (senders, receiver) = std::sync::mpsc::channel();
 
-    for c in configs.iter() {
+    for (thread_id, c) in configs.iter().enumerate() {
         let config = c.clone();
         let cid_map = cid_map.clone();
+        let sender = senders.clone();
+        let name = format!("vhu-vsock-cid-{}", c.get_guest_cid());
         let handle = thread::Builder::new()
-            .name(format!("vhu-vsock-cid-{}", c.get_guest_cid()))
-            .spawn(move || start_backend_server(config, cid_map))
+            .name(name.clone())
+            .spawn(move || {
+                let result =
+                    std::panic::catch_unwind(move || start_backend_server(config, cid_map));
+
+                // Notify the main thread that we are done.
+                sender.send(thread_id).unwrap();
+
+                result.map_err(|e| BackendError::ThreadPanic(name, e))?
+            })
             .unwrap();
-        handles.push(handle);
+        handles.insert(thread_id, handle);
     }
 
-    for handle in handles {
-        handle.join().unwrap()?;
+    while !handles.is_empty() {
+        let thread_id = receiver.recv().unwrap();
+        handles.remove(&thread_id).unwrap().join().unwrap()?;
```

Comment:
I just stumbled over this commit while rebasing old code. I wonder: Wouldn't setting `panic = "abort"` in profile config of the Cargo.toml do the same (without the extra complexity?).

Reply:
I don't know what you're referring to by "do the same" here, but `panic = "abort"` turns off unwinding the stack, which results in not running Drop in general. It's something you do when you don't care about cleaning up and/or you want a smaller binary size.

Reply:
> I don't know what you're referring to by "do the same" here, but `panic = "abort"` turns off unwinding the stack, which results in not running Drop in general. It's something you do when you don't care about cleaning up and/or you want a smaller binary size.

Right. It would not be exactly the same, but we are not running Drop for all the other threads in this case either, right? So it would not be too different. The OS should release all resources of the dying process. But of course it would be one setting for ALL threads. So if there are other threads involved, then that may be undesired.

I mostly wondered whether that was considered. But I am also not having particulary strong opinions here.

Reply:
I still don't understand what panic=abort has to do with this line though :thinking: Setting this would affect all panics, not just this line.


> So it would not be too different. The OS should release all resources of the dying process. But of course it would be one setting for ALL threads. So if there are other threads involved, then that may be undesired.

It's not about free()ing memory/memory leaks, it's about running the destructor logic.

Reply:
> I still don't understand what panic=abort has to do with this line though 🤔 Setting this would affect all panics, not just this line.

This line does a .join().unwrap(), which kills the main thread (taking all the child threads with it - without cleaning those up).

> Setting this would affect all panics, not just this line.

Sure. Though, I fail to see where that would be a problem.

> It's not about free()ing memory/memory leaks, it's about running the destructor logic.

The only difference is that without abort = true, you run the destructors of the crashing thread and those of the main thread. All other threads just vanish with the main thread gone.

Anyway, I was just wondering and thinking out loud. I see no need to respin this.

Reply:
Here more than the panic I'm interested in the possible error of any thread and get out. For the panic, yes maybe what you say would be enough, we could try that.

Reply:
Basically what was happening here was that if one thread ended with an error (no a panic), the main one would be left waiting for all the others that never ended. If it panicked instead, I think it panicked the main as well.

Reply:
> Basically what was happening here was that if one thread ended with an error (no a panic), the main one would be left waiting for all the others that never ended. If it panicked instead, I think it panicked the main as well.

Ah. I missed that part. There abort would not help of course. Sorry for the noise!